swarm/network/stream/delivery.go

```diff
@@ -194,26 +231,59 @@ func (d *Delivery) FindPeer(req *storage.Request) (*Peer, error) {
 
 		// skip peers that we have already tried
 		if req.SkipPeer(id.String()) {
+			skipped++
 			log.Trace("findpeer skip peer", "peer", id, "ref", req.Addr.String())
 			return true
 		}
 
-		// if origin is farther away from req.Addr and origin is not in our depth
-		prox := chunk.Proximity(req.Addr, d.kad.BaseAddr())
-		// proximity between the req.Addr and our base addr
-		if po < depth && prox >= depth {
+		// compare to MinBinSize == 2
+		if skipped >= 2 {
+			log.Trace("already skipped MinBinSize legit peers, no point continuing the search", "skipped", skipped)
+
+			err = fmt.Errorf("skipped MinBinSize legit peers, can't find chunk")
+			return false
+		}
+
+		// make sure we don't forward a request, when it is coming from our nearest neighbourhood
+		if po <= myPo && po <= originPo {
```

Comment:
This and the next condition I think are incorrect. Simply:

```golang
if myPO < depth { //  chunk is NOT within the neighbourhood
   if po <=  myPO { // always choose a peer strictly closer to chunk than us
       return false
  }
} else  { // chunk IS WITHIN neighbourhood
   if po < depth { // do not select  peer outside the neighbourhood. But allows peers futher from the chunk then us
       return false
   }
}
```

Reply:
@zelig I will try this and see if it works - but it seems to be missing the part that STOPS requests being forwarded within the neighbourhood.

The second case: if the chunk is within the neighbourhood and the request came from a peer in the neighbourhood, you should NOT forward at all - otherwise you end up with loops. I am not sure if this is enough to prevent loops altogether, but if this logic is not there, then we definitely have loops.

Reply:
@zelig I tried this solution, and as I thought it doesn't prevent loops in the nearest neighbourhood.

Reply:
This condition that you think is incorrect on the other hand didn't result in loops.

Reply:
Changing to:
```
		if myPo < depth { //  chunk is NOT within the neighbourhood
			if po <= myPo { // always choose a peer strictly closer to chunk than us
				return false
			}
		} else { // chunk IS WITHIN neighbourhood
			if po < depth { // do not select  peer outside the neighbourhood. But allows peers futher from the chunk then us
				return false
			} else if po <= originPo { // avoid loop in neighbourhood, so not forward when a request comes from the neighbourhood
				return false
			}
		}
```
Let's see if this works when we hit asymmetric relationships in terms of nearest neighbourhood.